lib/Basics/coroutine/thread_registry.h

```diff
@@ -0,0 +1,136 @@
+    promise_head.store(promise, std::memory_order_release);
+  }
+
+  /**
+     Executes a function on each promise in the registry.
+
+     Can be called from any thread. It makes sure that all
+     items stay valid during iteration (i.e. are not deleted in the meantime).
+   */
+  template<typename F>
+  requires requires(F f, PromiseInList* promise) { {f(promise)}; }
+  auto for_promise(F&& function) noexcept -> void {
+    auto guard = std::lock_guard(mutex);
+    // (2) - this load synchronizes with store in (1)
+    for (auto current = promise_head.load(std::memory_order_acquire);
+         current != nullptr;
+         // (5) - this load synchronizes with store in (3) and (4)
+         current = current->next.load(std::memory_order_acquire)) {
+      function(current);
+    }
+  }
+
+  /**
+     Marks a promise in the registry for deletion.
+
+     Can be called from any thread. The promise needs to be included in
+     the registry list, crashes otherwise.
+   */
+  auto mark_for_deletion(PromiseInList* promise) noexcept -> void {
+    // makes sure that promise is really in this list
+    ADB_PROD_ASSERT(promise->registry == this);
+    // (1) - this load synchronizes with compare_exchange_weak in (2)
+    auto current_head = free_head.load(std::memory_order_acquire);
+    do {
+      promise->next_to_free.store(current_head, std::memory_order_relaxed);
+      // (2) - this compare_exchange_weak synchronizes with load in (1)
+    } while (not free_head.compare_exchange_weak(current_head, promise,
+                                                 std::memory_order_release,
+                                                 std::memory_order_relaxed));
+  }
+
+  /**
+     Deletes all promises that are marked for deletion.
+
+     Can only be called on the owning thread, crashes otheriwse.
+   */
+  auto garbage_collect() noexcept -> void {
+    ADB_PROD_ASSERT(std::this_thread::get_id() == owning_thread);
+    PromiseInList* head;
+    {
+      auto guard = std::lock_guard(mutex);
+      // (7) - this exchange synchronizes with load in (1) and store in (2)
+      head = free_head.exchange(nullptr, std::memory_order_acq_rel);
+    }
+    for (auto current = head; current != nullptr;) {
+      auto next = current->next_to_free.load(std::memory_order_relaxed);
+      remove(current);
+      current->destroy();
+      current = next;
+    }
+  }
+
+  const std::thread::id owning_thread = std::this_thread::get_id();
+  std::atomic<PromiseInList*> free_head = nullptr;
+  std::atomic<PromiseInList*> promise_head = nullptr;
+  std::mutex mutex;
+
+ private:
+  /**
+     Removes the promise from the registry.
+
+     Caller needs to make sure that the given promise is part of this registry
+     (which also means that this function should only be called on the owning
+     thread.)
+   */
+  auto remove(PromiseInList* promise) -> void {
+    auto next = promise->next.load(std::memory_order_relaxed);
```

Comment:
```suggestion
  auto remove(PromiseInList* promise) -> void {
    ADB_PROD_ASSERT(std::this_thread::get_id() == owning_thread);
    auto next = promise->next.load(std::memory_order_relaxed);
```

Reply:
This function is only called in the garbage_collect fct which does this check, so if I would add it here it would be done twice. I extracted remove into a function for better readability. I would prefer leaving it as it is, it is private, so it cannot be called from outside accidentally.